File: wallet-web/app/src/api/swapsApi.ts
import { apiRequest } from './httpClient';
import { assertOnline } from '../connectivity/assertOnline';
import type { ActionProposal } from './paymentsApi';

// wallet-backend's swaps component is a generic DEX router-call builder,
// not an abstracted "swap A for B" - the caller supplies the router
// contract, its ABI, and which method/args to call (PLAN.md §2:
// "the project configures which router it wants, this code doesn't pick
// for it"). wallet-web's own swap page (PLAN.md §9 Phase 11) is
// responsible for knowing a specific router's ABI/method shape for
// whichever DEX it targets.
export interface BuildSwapInput {
  routerAddress: string;
  routerAbi: string;
  method: string;
  args?: unknown[];
  valueWei?: string;
}

export async function buildSwap(walletAddress: string, input: BuildSwapInput): Promise<ActionProposal> {
  await assertOnline();
  return apiRequest<ActionProposal>('/v1/swaps/build', { method: 'POST', walletAddress, body: input });
}

export function submitSwap(walletAddress: string, actionId: number, signature: string): Promise<{ hash: string }> {
  return apiRequest<{ hash: string }>('/v1/swaps/submit', {
    method: 'POST',
    walletAddress,
    body: { actionId, signature },
  });
}

File: wallet-mobile/app/lib/api/swaps_api.dart
// Mirrors wallet-web's api/swapsApi.ts. wallet-backend's swaps component
// is a generic DEX router-call builder, not an abstracted "swap A for B"
// - the caller supplies the router contract, its ABI, and which
// method/args to call. This app's swap page is responsible for knowing a
// specific router's ABI/method shape for whichever DEX it targets.
import 'http_client.dart';
import 'payments_api.dart' show ActionProposal;

class SwapsApi {
  SwapsApi(this._client);
  final ApiClient _client;

  Future<ActionProposal> buildSwap(
    String walletAddress, {
    required String routerAddress,
    required String routerAbi,
    required String method,
    List<dynamic>? args,
    String? valueWei,
  }) {
    return _client.request(
      '/v1/swaps/build',
      method: 'POST',
      walletAddress: walletAddress,
      body: {
        'routerAddress': routerAddress,
        'routerAbi': routerAbi,
        'method': method,
        'args': args ?? [],
        if (valueWei != null) 'valueWei': valueWei,
      },
      decode: (json) => ActionProposal.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<String> submitSwap(String walletAddress, int actionId, String signature) {
    return _client.request(
      '/v1/swaps/submit',
      method: 'POST',
      walletAddress: walletAddress,
      body: {'actionId': actionId, 'signature': signature},
      decode: (json) => (json as Map<String, dynamic>)['hash'] as String,
    );
  }
}
